File: 240529/트리 노드 제거/remove-tree-node.py
from collections import deque
import sys

input = sys.stdin.readline

n = int(input())

parents = list(map(int, input().split()))

delete = int(input())

G = [[] for _ in range(n)]

for child, parent in enumerate(parents):
    if parent == -1:
        start = child
        continue
    elif parent == delete or child == delete:
        continue
    G[parent].append(child)

visited = [False for _ in range(n)]

def bfs(s):
    cnt = 0
    visited[s] = True
    q = deque()
    q.append(s)
    while(q):
        x = q.popleft()
        if len(G[x]) == 0:
            cnt = cnt + 1
        else:
            for child in G[x]:
                if not visited[child]:
                    visited[child] = True
                    q.append(child)
    return cnt
print(bfs(start))

File: 240529/트리 노드 제거/remove-tree-node.cpp
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

int n, del, start;
int parents[50];
vector <int> G[50];

int bfs(int s)
{
    if (s == del)
        return 0;
    int cnt = 0;
    queue <int> q;
    q.push(s);
    while(!q.empty())
    {
        int x = q.front();
        q.pop();
        bool is_leaf = true;
        for (int i = 0; i < G[x].size(); i++)
        {
            int child = G[x][i];
            if (child == del)
                continue;
            q.push(child);
            is_leaf = false;
        }
        if (is_leaf)
            cnt++;
    }
    return (cnt);
}

int main() {
    cin >> n;
    for (int i = 0; i < n; i++)
        cin >> parents[i];
    cin >> del;

    for (int i = 0; i < n; i++)
    {
        int child = i, parent = parents[i];
        if (parent == -1)
        {
            start = child;
            continue;
        }
        G[parent].push_back(child);
    }
    cout << bfs(start);

    return 0;
}